msp/msp_exp_handler.c

```diff
@@ -167,213 +167,149 @@ static uint16_t table2[129] = {////table 1 with compression  1, 1,2,4,8,16,32,64
-	                        + (histo_data[k + i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4] & 0xFFFF);
-	 		}
-			bin = bin + (histo_data[k + i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4] & 0xFFFF);
-				bin = bin/bin_size;
-		}
-		else if (carry_over == 1 && bin_size%2 == 1) {// carry over with uneven bin size
-			carry_over = 1;
-			carry_over = 0;
-			k = (*(table+j)-1)/2;
-			bin = histo_data[k + i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4]>>16 & 0xFFFF;
-			if (bin_size != 1) {
-				for (k=(*(table+j)+1)/2; k<*(table+j+1)/2; k++) {
-					bin = bin + (histo_data[k + i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4]>>16 & 0xFFFF)
-	                    	  + (histo_data[k + i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4] & 0xFFFF);
-	 		    }
-				bin = bin/bin_size;
-			}
-		}
-			send_data_payload[j*2 + 1 + len] = bin & 0xFF; //store bin
-			send_data_payload[j*2 + len] = bin>>8 & 0xFF;
-			bin = 0;
-	}
-	return len + num_bins*2;//2 bytes per bin
-
-}
-
-static unsigned long prep_payload_data(uint8_t *bin_config){
-	/*Used to compress histogram data into bins of chosen size, puts the compressed bins into
-	 * send_data_payload. returns the variable len = byte length of send_data_payload*/
-    unsigned long i, len;
+static unsigned long prep_payload_data(uint8_t *bin_config)
+{
+    unsigned long i, j, k, len;
     uint16_t num_bins;
+    uint8_t bin_size;
+    uint32_t bin;
+
+    uint16_t send_index; // start index used for send data payload
+    uint16_t data_index; // Index used to access data form histo_data
+    uint16_t start_index; // first index in histo_data
 
 	uint32_t *histo_data = (uint32_t *)HISTO_ADDR;
 	/* histogram header into send_data_payload */
 	for (i=0; i<HISTO_HDR_NUM_BYTES/4; i++) {
-		send_data_payload[i*4 + 1] = histo_data[i] & 0xFF;
-		send_data_payload[i*4 + 0] = histo_data[i]>>8 & 0xFF;
-		send_data_payload[i*4 + 3] = histo_data[i]>>16 & 0xFF;
-		send_data_payload[i*4 + 2] = histo_data[i]>>24 & 0xFF;
+		send_index = i*4;
+		send_data_payload[send_index + 1] = histo_data[i] & 0xFF;
+		send_data_payload[send_index + 0] = histo_data[i]>>8 & 0xFF;
+		send_data_payload[send_index + 3] = histo_data[i]>>16 & 0xFF;
+		send_data_payload[send_index + 2] = histo_data[i]>>24 & 0xFF;
 	}
 	len = HISTO_HDR_NUM_BYTES;
+
 	for (i=0; i<6; i++) {
 		/* No re-binning */
+		start_index = i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4;
+
 		if (bin_config[i] == 0) {
-			len = no_binning(len, i, histo_data);
+			for (j=0; j<HISTO_NUM_BINS_GW/2; j++) {
+				send_index = j*4 + len;
+				data_index = j + start_index;
+				send_data_payload[send_index + 1] = histo_data[data_index] & 0xFF;
+				send_data_payload[send_index] = histo_data[data_index]>>8 & 0xFF;
+				send_data_payload[send_index + 3] = histo_data[data_index]>>16 & 0xFF;
+				send_data_payload[send_index + 2] = histo_data[data_index]>>24 & 0xFF;
+			}
+			len = len + HISTO_NUM_BINS_GW*2;
     	}
+
 		/* Re-binning with equal interval bins */
 		else if (bin_config[i] < 7) {
```

Comment:
Move this `else if` to the line above, right after the closing brace, i.e.:

`} else if (...) {`